### ISSUE TITLE
Running into issues cloning images with cloud-init drives attached

### ISSUE THREAD

**Issue by n8rade** [Telmate/terraform-provider-proxmox#704]
When I attempt to clone certain images it appears to be trying to set the scsi0 drive as the cloud-init drive. This only started occurring after updating to the latest provider version 2.9.13.

Error seen:
~~~
scsi0 - cloud-init drive is already attached at 'ide0'
~~~

I can see it trying to assign the cloudinit named drive to the SCSI device when this is incorrect:
~~~
-scsi0 local:125/vm-125-cloudinit.qcow2,size=50G,backup=0,ssd=1
~~~

For the above error the `vm-125-cloudinit.qcow2` device already exists on ide0 as the error mentions. ide0 is the cloud-init device.

- This behavior isn't consistent. I had a RHEL 8 image that was working for a few days after upgrading the provider version and then spontaneously stopped work.
- I can't explain this behavior at all. 
- I don't believe it's a config issue, but am happy to share those if it may be helpful.
- This is not consistent to all my images.
- When I first ran into the issue on a RHEL 7.9 image I recreated the image with packer and the error went away. This did not work for the RHEL 8 image I'm running into this for now.

Attached is the debug log I gathered for a bad creation [terraform-plugin-proxmox.log](https://github.com/Telmate/terraform-provider-proxmox/files/10843110/terraform-plugin-proxmox.log). 

Here is a successful configure sent through to a RHEL 8.5 based VM I mentioned earlier showing the desired/intended behavior:
~~~
-scsi0 local:125/vm-125-disk-0.qcow2,size=50G,backup=0,ssd=1
~~~

If this issue lies elsewhere I'd be happy to go to that repo/source for more help, I just have no idea where to look at this point.

Thank you!

**Comment by D-J-See** [comment]
I had the same issue yesterday after updating to provider version 2.9.13 using a packer created AlamaLinux 9.1 image. When deploying a k8s workspace with 4 hosts, of the ~10 attempts, between 1 and 3 of the VMs had the issue, strangely with at least 1 VM per attempt not having it. Worked perfectly first time when I went back to provider version 2.9.11.
I was running:

    ProxMox 7.3-6.
    Terraform v1.3.9
    telmate/proxmox v2.9.13

**Comment by lsampaioweb** [comment]
I am having the same problem when cloning a template created by Packer (cloud-init = true). I found a way that works, not the best thing, but it works. I hope you guys can find a better solution.

What I did was manually remove the cloud-init drive from ide0 and add a new one on ide1, ide2 or ide3, now it works. But this manual action is something that I don't want to keep doing.

Just like @n8rade said, I did not have this problem before.

Can you guys help me? 

My homelab:
ProxMox 7.3-6.
Terraform v1.4.0.
telmate/proxmox v2.9.13.

**Comment by kayvenm** [comment]
I am also having the same problem so I spent a few hours investigating the problem and from what I can tell it's caused by the NewConfigQemuFromApi function in the [proxmox-api-go](https://github.com/Telmate/proxmox-api-go) project. At the end of of the disk loop, it's adding the disk to an array using the diskID extracted from the diskname (scsi**0**, ide**0**) (https://github.com/Telmate/proxmox-api-go/blob/ad6090b8284272279e238bcaafadaa0e28536d7c/proxmox/config_qemu.go#L796) so having both a scsi0 and ide0 will overwrite each other. This also confirms what @lsampaioweb found as a work around. I'm not familiar enough with proxmox-api-go to fix the problem. Instead of having a common array for all disk types, having a separate one for each type (ide, scsi, ...) would fix the problem but it breaks the current API. Appending all the disks to the array instead of writing at a specific index would also fix it but I have no idea what else it might break. Hopefully somebody with more knowledge can chime in.

**Comment by kctang** [comment]
I encountered similar issues. I managed to overcome this issue by specifying the disk more explicitly.

TL;DR - if you specify the `disk` block explicitly (and correctly), it works. It needs to be exactly how your template's disk is defined.

As I am not very experienced in Terraform myself, instead of providing the solution, I will just provide these hints/steps I took to overcome my specific situation. Hope it helps others.

My terraform file looks like this:

```
resource "proxmox_vm_qemu" "vm_template" {
  for_each = var.vm_list

  clone  = "ubuntu-22.04.2-template"
  vmid   = each.value.vmid
  name   = each.key
  desc   = each.value.desc
  cores  = each.value.cores
  memory = each.value.memory

  target_node = "pve"
  agent       = 1
  cpu         = "host"
  qemu_os     = "l26"

  scsihw    = "virtio-scsi-pci"
  os_type   = "cloud-init"
  ipconfig0 = each.value.ipconfig0

  disk {
    backup  = true
    cache   = "none"
    file    = "vm-${each.value.vmid}-disk-0"
    format  = "raw"
    size    = each.value.disk
    storage = "local-lvm"
    type    = "virtio"
    volume  = "local-lvm:vm-${each.value.vmid}-disk-0"
  }

  lifecycle {
    ignore_changes = [
      network,
    ]
  }
}
```

- My template is a cloud-init based template created via packer (for proxmox).
- It creates multiple VMs using `for_each` based on `var.vm_list`. However, that's the really the concern here.
- If you **specify details in the `disk` block correctly**, terraform will apply correctly. I suspect `backup` and `cache` to be optional but did not try removing them.
- How I figure out the "correct" disk block? First, I apply terraform **without** the disk block (just remove the entire block). That would generate my VM based on the cloned template's disk configuration. Review the disk's configuration of the applied state and use the values in your terraform configuration.

Hope this helps someone struggling with similar issues.

**Comment by n8rade** [comment]
(Quote snipped some)
> TL;DR - if you specify the `disk` block explicitly (and correctly), it works. It needs to be exactly how your template's disk is defined.
> * If you **specify details in the `disk` block correctly**, terraform will apply correctly. I suspect `backup` and `cache` to be optional but did not try removing them.
> * How I figure out the "correct" disk block? First, I apply terraform **without** the disk block (just remove the entire block). That would generate my VM based on the cloned template's disk configuration. Review the disk's configuration of the applied state and use the values in your terraform configuration.

First I did want to state thank you for the workaround! I haven't gotten around to testing it yet but it seems promising!

According to the documentation it would appear that there are only a few required arguments for the disk block: https://registry.terraform.io/providers/Telmate/proxmox/latest/docs/resources/vm_qemu#disk-block

Maybe I'm misinterpreting your use of the word "correctly" incorrectly here, but I would assume that defining only the bare minimum required blocks would be the correct way. Anything more than that should be optional unless otherwise specified.

Here's how I'm defining my disks:
~~~
  disk {
    size         = "${var.vm_disk_size}"
    storage      = "${var.vm_storage}"
    type         = "${var.vm_disk_type}"
    ssd		 = "${var.vm_ssd}"
    backup = "${var.vm_backup}"
~~~

Based on the error seen and how you worked around this, I would assume that adding in the `file` and `volume` parameters should allow this to work by explicitly defining them. But both of them state roughly the same thing in the docs:
~~~
file	str		The filename portion of the path to the drive’s backing volume. You shouldn't need to specify this, use the storage parameter instead.
volume	str		The full path to the drive’s backing volume including the storage pool name. You shouldn't need to specify this, use the storage parameter instead.
~~~

Based on this I would hope a fix would still be done and this issue not looked over! But very happy to have a workaround!

**Comment by djh00t** [comment]
I have also hit this issue, the only way I could get around it was to create a 4M dummy device on virtio0 and make virtio1 the default boot drive.  Its ugly but it works for now.


![Screenshot 2023-04-04 at 3 12 21 pm](https://user-images.githubusercontent.com/13357567/229693616-77876541-9b38-47ff-b3f4-404be95296ac.png)

**Comment by hkbakke** [comment]
I suspect the issue outlined in https://github.com/Telmate/terraform-provider-proxmox/issues/704#issuecomment-1474198878 is also causing other nasty issues, like extra disk images being added at random terraform plan/apply runs if ide0 and scsi0 exists on the same VM. Hopefully this will get a proper fix as the workarounds coverede here does nothing about the underlying problems.

**Comment by Sm0rezDev** [comment]
I have also stumble across this dreaded issue, It only happen when you explicitly declare a disk block to a template configured with cloud-init drive. Gonna be a better workaround when you need to increase the cloned template disk size without having to deal with a nasty bug.

**Comment by n8rade** [comment]
I can confirm that the issue still exists with plugin version 2.9.14:
~~~
│ Error: scsi0 - cloud-init drive is already attached at 'ide0'
│ 
│   with module.full-work-creation.module.vm-work-creation["test1"].proxmox_vm_qemu.vm-work-creation,
│   on .terraform/modules/full-work-creation.vm-work-creation/main.tf line 11, in resource "proxmox_vm_qemu" "vm-work-creation":
│   11: resource "proxmox_vm_qemu" "vm-work-creation" {
│ 
╵
╷
│ Error: scsi0 - cloud-init drive is already attached at 'ide0'
│ 
│   with module.full-work-creation.module.vm-work-creation["pah"].proxmox_vm_qemu.vm-work-creation,
│   on .terraform/modules/full-work-creation.vm-work-creation/main.tf line 11, in resource "proxmox_vm_qemu" "vm-work-creation":
│   11: resource "proxmox_vm_qemu" "vm-work-creation" {
│ 
╵
╷
│ Error: scsi0 - cloud-init drive is already attached at 'ide0'
│ 
│   with module.full-work-creation.module.vm-work-creation["controller"].proxmox_vm_qemu.vm-work-creation,
│   on .terraform/modules/full-work-creation.vm-work-creation/main.tf line 11, in resource "proxmox_vm_qemu" "vm-work-creation":
│   11: resource "proxmox_vm_qemu" "vm-work-creation" {
~~~

And here's showing that 2.9.14 is in use:
~~~
# grep Telmate providers.tf -A1
      source = "Telmate/proxmox"
      version = ">= 2.9.14"
# ll .terraform/providers/registry.terraform.io/telmate/proxmox/2.9.14/
total 0
drwxr-xr-x. 2 root root 80 Jun 10 09:10 linux_amd64
~~~

Debug logs can be found in original post. Disk configuration has also been provided, but if other configuration is needed please let me know.

**Comment by hallowen** [comment]
There is a workaround for this. When disk i configured in terraform, make sure to add it in a different slot

```
  disk {
    size    = "500G"
    type    = "virtio"
    storage = "data-sata-ssd"
    format  = "raw"
    slot = 2
  }
```

**Comment by nkunaev** [comment]
So, I find a few ways to fix it.
1st:
Detach ide0 with cloud-init drive from base template in Proxmox
![изображение](https://github.com/Telmate/terraform-provider-proxmox/assets/116570149/af269962-4c2f-44d9-8943-041c8709bc3e)  

2nd:
If you're creating template via Packer, add the following flags to _source "proxmox-iso"_ block (provider hashicorp/proxmox 1.1.3):  

```hcp
unmount_iso = true  
cloud_init = false
```
to create template without cloud-init drive

Now I'm creating VM's without any problems.

**Comment by nkunaev** [comment]
So, look's like found few ways to fix it.
1. Remove cloud-init drive from template in Proxmox
![изображение](https://github.com/Telmate/terraform-provider-proxmox/assets/116570149/dbca5408-53da-47a8-982f-522ed88a9022)

1.1. If you're creating VM via Packer (provider hashicorp/proxmox, ver. 1.1.3), in block **source "proxmox-iso"** add the following flags:
```
unmount_iso = true
cloud_init  = false
```
It will unmount cloud-init drive before convert VM into template

2. If you're setting custom cloud-init parameters in Terraform such as "ciuser", "cipassword", "sshkeys" and etc., you should add the following:
```
resource "proxmox_vm_qemu" "cloudinit_vm" {
.....
os_type = "cloud-init"  
cloudinit_cdrom_storage = "local_lvm" (or another node storage name)
...
}
```
This parameter add cloud-init drive to your VM and it will work.

**Comment by hestiahacker** [comment]
Oof, this is the third bug (so far) that has bitten me from 2.9.13 (and later)!

Workaround report:
1. [Specifying the disk parameters](https://github.com/Telmate/terraform-provider-proxmox/issues/704#issuecomment-1475302632) (backup, size, type, cache and storage) did not work around the issue
2. [Specifying cloudinit_cdrom_storage](https://github.com/Telmate/terraform-provider-proxmox/issues/704#issuecomment-1645291953) did not work around this issue
3. [slot = 2](https://github.com/Telmate/terraform-provider-proxmox/issues/704#issuecomment-1628961851) did not work
4. [revert to 2.9.11](https://gitlab.com/acidpizza-stuff/infra/terraform-modules/proxmox-vm-terraform-module) **worked**
5. manually [removing the CloudInit drive from the template](https://github.com/Telmate/terraform-provider-proxmox/issues/704#issuecomment-1645291953) via the Proxmox UI **only works if you don't need cloudinit** (e.g. to set a static IP for the VM). However, re-adding it on IDE-2
6. [rebuilding the template with packer](https://github.com/Telmate/terraform-provider-proxmox/issues/704#issuecomment-1645291953) and unmount_iso=true and cloud_init=false failed in the same way as removing it manually

I'll prioritize setting up a test environment so I can make code changes to the provider and try to get all of these bugs fixed which were introduced in 2.9.13. I've been locked on 2.9.11 since Feburary because when I initially updated to 2.9.13.

**Comment by hestiahacker** [comment]
On number 5 of my list above I should have said re-adding it on IDE-2 also **worked**. Apparently I got distracted part way through while waiting for it to complete.

**Comment by FlexibleToast** [comment]
I have been having this issue as well. The solution I finally found to work was similar to hestiahacker's #5. I removed the cloud-init drive from the template. The difference or at least what wasn't stated was that in order to add it back I just specified `cloudinit_cdrom_storage = "local-zfs"`. This created a new cloud-init drive on the cloned VM that then assigned its own slot. This also retained the cloud-init settings that were being used by the template before I removed the drive.

**Comment by michael-akkermans** [comment]
For me deleting the cloud-init drive from the template  and re-adding it to the template on IDE-1 fixed it in my environment. Thanks for the advice here.

**Comment by chrishas35** [comment]
> I have been having this issue as well. The solution I finally found to work was similar to hestiahacker's #5. I removed the cloud-init drive from the template. The difference or at least what wasn't stated was that in order to add it back I just specified `cloudinit_cdrom_storage = "local-zfs"`. This created a new cloud-init drive on the cloned VM that then assigned its own slot. This also retained the cloud-init settings that were being used by the template before I removed the drive.

This was the trick for me as well.

**Comment by TheBelgarion** [comment]
that bugged me too, somehow its not always - would be nice if there was a fix for that.

**Comment by willifehler** [comment]
Any update on this issue?

**Comment by hestiahacker** [comment]
Hey everyone, I just wanted to let you all know that I haven't forgotten about this ticket. To follow up on my [last](https://github.com/Telmate/terraform-provider-proxmox/issues/704#issuecomment-1646519281) two [comments](https://github.com/Telmate/terraform-provider-proxmox/issues/704#issuecomment-1646713406): I now have a test Proxmox server, with test users and it's an environment that I can really tinker with and not have to worry about bringing down production.

I have been able to replicate the issue in my test environment. My terraform works fine with 2.9.11 and gives a `Error: virtio0 - cloud-init drive is already attached at 'ide0'` error when I switch to 2.9.14. Same versions of Proxmox and Terraform, same terraform files. The only thing that changed was the provider.

Next I'm going to try building the latest dev branch of everything and if that doesn't work, I'll start trying to track down exactly what's causing this and fix the bug.

If anyone has an easy way to see a debug log of every request being made to Proxmox (ideally with a stacktrace), that might help me get this resolved more quickly. Feel free the send the links my way!

**Comment by n8rade** [comment]
Hey @hestiahacker, hopefully you can nail this one down! Here's how I enabled debug logs when I originally reported the issue: [https://registry.terraform.io/providers/Telmate/proxmox/latest/docs#pm_log_enable](https://registry.terraform.io/providers/Telmate/proxmox/latest/docs#pm_log_enable)
There's an example debug log in my original report if you wanted to see what that looked like.

There's likely more logging available, but that's the most I could find related to the plugin at least.

**Comment by hestiahacker** [comment]
The TL;DR is that I believe the issue was in a library that we depend on, they fixed it upstream, we picked up the fix in 1fe6cf7a0184371bce277737df5134e61e106ea0, and this should be fixed in the next release. In the meantime, if you really need a workaround and v2.9.14 isn't working, checkout 4c0c8fc37ed3f6316ad9392a46a0cf28c4654a11 (which is the latest working commit) and compile that.

I'm afraid have some frustrating results to share. I'm no longer able to reproduce the issue in v2.9.14. Here's the rundown of where I can and can not reproduce the problem:

- v2.9.14 works fine now (WTF?)
- v2.9.11 still works fine too
- e51e787c3a792176a6401c2db1b15cb2a30449e0 (latest at time of writing) causes the VM to not find the primary boot disk, and thus not boot
- a8675d3967710bab4ac08fad9dbc05eed3ae2c58 compiled, but failed because my VM name had dots and dots were not allowed in this version
- f6a718cd1e950ed74e4fff639a4ab4270e5a3707 didn't compile
- bf66e7ebeaf8d0e986f69ec64229dbd1a02d3b6e didn't compile
- 905bcd9b55aa7a60c0254e617db2096d03acf276 didn't compile
- d02bb5dba45bdb8e65ed58fccaa885fd6432e6fa didn't compile
- 568cb331526d2fb65dd5edb3aab9275c5eabc7cf didn't compile
- 704db887ccbe636e2d97bafbc969eee56f99c13a worked fine
- 4c0c8fc37ed3f6316ad9392a46a0cf28c4654a11 **not able to reproduce the problem**, everything worked fine!
- 1fe6cf7a0184371bce277737df5134e61e106ea0 **able to reproduce the problem**
- bd28855b7c34d9d6913f4ac22c1933610a085cc2 able to reproduce the problem

Again, I didn't change anything other than the namespace/version of the provider in my terraform config. Same proxmox server, same version of Proxmox, same terraform files.

The [commit](1fe6cf7a0184371bce277737df5134e61e106ea0) where it went from being a reproducible issue to one that I can't reproduce is that change is something from the iso9660 library. It seems plausible that the bug was there since the cloud-init feature is implemented by putting configuration data onto a CD-ROM (iso9660).

The issue with the latest commit (e51e787c3a792176a6401c2db1b15cb2a30449e0) appears to be because it's adding two hard drives (disk-0, disk-1) instead of one (disk-0) and disk-0 is not connected whereas disk-1 is connected to virtio0). It is being tracked as https://github.com/Telmate/terraform-provider-proxmox/issues/887 and I've made a comment over there to hopefully help out with that one.

On a lighter note, I have made some documentation on how developers who are new to this project can get started. I submitted a [merge request](https://github.com/Telmate/terraform-provider-proxmox/pull/888) in hopes that it'll get included in this repo. I'm sure I'll use it next time I go to debug something in the provider and hopefully it'll be accepted so others will find it and will have the tools and knowledge to investigate issues further.

**Comment by n8rade** [comment]
I've found the issue to be intermittent. Roll of the dice for me to see if I'll hit it. I'm assuming you might be running into that intermittency now?

I appear to have been utilizing 2.9.14 since June 10th:
~~~
# ll -d .terraform/providers/registry.terraform.io/telmate/proxmox/2.9.14
drwxr-xr-x. 3 root root 25 Jun 10  2023 .terraform/providers/registry.terraform.io/telmate/proxmox/2.9.14
~~~
~~~
# ll providers.tf
-rw-r--r--. 1 root root 714 Jun 10  2023 providers.tf
~~~
And do utilize it:
~~~
# grep "Telmate/proxmox" providers.tf -A1 -B1
    proxmox = {
      source = "Telmate/proxmox"
      version = ">= 2.9.14"
~~~

And with my first `apply` I can confirm I still run into it:
~~~

│ Error: scsi0 - cloud-init drive is already attached at 'ide0'
│
│   with module.full-work-creation.module.vm-work-creation["eda"].proxmox_vm_qemu.vm-work-creation,
│   on .terraform/modules/full-work-creation.vm-work-creation/main.tf line 11, in resource "proxmox_vm_qemu" "vm-work-creation":
│   11: resource "proxmox_vm_qemu" "vm-work-creation" {
│
╵
╷
│ Error: scsi0 - cloud-init drive is already attached at 'ide0'
│
│   with module.full-work-creation.module.vm-work-creation["eda2"].proxmox_vm_qemu.vm-work-creation,
│   on .terraform/modules/full-work-creation.vm-work-creation/main.tf line 11, in resource "proxmox_vm_qemu" "vm-work-creation":
│   11: resource "proxmox_vm_qemu" "vm-work-creation" {
│
╵
~~~

I've been resistant to using any of the workarounds here for one reason or another. But with the above `apply` 2 out of 3 of the VM's failed.

PVE 7.4-17
Terraform v1.6.6

**Comment by hestiahacker** [comment]
Ah yes, the cloud-init attachment appears to be an intermittent issue, as just commented by [n8rade](https://github.com/n8rade) and also mentioned in #876. Once I realized this, I tried redeploying with v2.9.14 a few more times and I was able to reproduce this issue with that version. So that explains why that version mysteriously started working today. It was just a fluke.

It also meant that I felt compelled to retest 4c0c8fc37ed3f6316ad9392a46a0cf28c4654a11 repeatedly to see if it consistently works. It does not. :sob:

I think I'm going to turn my attention to #887 because that's preventing me from testing using the latest commit. I'll take a look at the fix that @TheGameProfi made which was mentioned in [a comment](https://github.com/Telmate/terraform-provider-proxmox/issues/863#issuecomment-1826285173) on that ticket. Maybe I can snag the code and make a merge request with their fix.

I think this is likely going to quickly escalate into me making changes to the `proxmox-api-go` repo. Hopefully I'll figure out how to build from a local directory instead of constantly having to commit, push, and update `go.sum` with a version/date/commitHash on each change before I can re-build the provider.

**Comment by TheGameProfi** [comment]
@hestiahacker I changed in my version the use of the storage, to include strings and floats.
And fixed some things from your newest merges.

For the use of local module you can use `replace()`

```
module github.com/userName/mainModule

require "github.com/userName/otherModule" v0.0.0
replace "github.com/userName/otherModule" v0.0.0 => "local physical path to the otherModule"
```
 https://stackoverflow.com/questions/52026284/accessing-local-packages-within-a-go-module-go-1-11

**Comment by hestiahacker** [comment]
Thank you @TheGameProfi for pointing out the `replace()` directive. I've updated the documentation and submitted a merge request.

I also have been able to get TinyBlardon's branch that [fixes #887](https://github.com/Telmate/terraform-provider-proxmox/issues/887#issuecomment-1876165346) to apply cleanly and have submitted a [merge request](https://github.com/Telmate/terraform-provider-proxmox/pull/889) for that.

I used that branch to try repeatedly deploying my VM to see if I can reproduce the issue with cloud-init drives not being attached. So far I've done 9 deploys in a row without hitting the issue. Previously I was hitting it at least once every 3 deploys.

And just to make sure I wasn't going crazy, I switched back to the 2.9.14 release and was able to reproduce the issue after 3 deploys. So I'm pretty confident that this issue really will be resolved with this new code.

**Comment by n8rade** [comment]
3.0.1-rc1 has been good to me so far in regards to this issue, so thank you @hestiahacker for testing early on! Will do more testing at final release as well.

**Comment by hestiahacker** [comment]
Wow, it sure would have been great if I had come up with and posted a minimal test case that could (intermittently) reproduce this issue. I have tested [this small example](https://github.com/Telmate/terraform-provider-proxmox/issues/887#issuecomment-1876176138), which does have a cloud-init drive, using [v3.0.1-rc1](https://github.com/Telmate/terraform-provider-proxmox/discussions/910) and it's deployed a few times in a row without any issue. So I am hopeful that this new release will take care of the issue and we'll be able to close this ticket.

**Comment by GitHubGeek** [comment]
Hi, I've just started using the provider v3.0.1 with PVE 8.1, and came across a small issue. The provider seems to be creating 2 `cdrom` devices instead of 1.

I created a VM template by importing `debian-12-genericcloud-amd64.qcow2` - Cloud-init isn't configured. Template has only one disk `scsi0`. Then, I defined a VM in Terraform to clone from the VM template.

The cloned VM somehow ended up with 2 cdrom, `ide2` and `ide3`. `ide2` has no media, `ide3` has the cloudinit disc. The VM works well by setting `boot = "order=scsi0;ide2;ide3"`. However, I'd like to know how to avoid creating the redundant cdrom device in the future. Help appreciated!

Tested cloning the VM template in PVE web UI - No cdrom device in the clone.

My Terraform config:

```hcl
resource "proxmox_vm_qemu" "vm1" {
  vmid        = 123
  target_node = "pve"
  name        = "vm1"
  desc        = "My VM"

  # The template name to clone this vm from
  clone      = "template-debian-12-cloud-init"
  full_clone = true
  qemu_os    = "l26"

  # Activate QEMU agent for this VM
  agent = 1

  os_type = "cloud-init"
  cores   = 4
  sockets = 1
  vcpus   = 0
  cpu     = "host"
  memory  = 8192
  balloon = 512

  scsihw = "virtio-scsi-pci"

  # Setup the disk
  disks {
    scsi {
      scsi0 {
        disk {
          storage = "local-lvm"
          size    = 384
          discard = true
          backup  = true
        }
      }
    }
  }

  boot                    = "order=scsi0;ide2;ide3"
  cloudinit_cdrom_storage = "local-lvm"
  cicustom                = "user=snips:snippets/userdata-debian-12.yaml"
  ipconfig0               = "ip=dhcp,ip6=auto"
}

```

**Comment by hestiahacker** [comment]
I'm unable to reproduce this issue with v3.0.1-rc1 of the provider and v7.4-17 of Proxmox.

My test setup is slightly different in that I am cloning a VM that already has 2 CD-ROM drives: ide0 (cloudinit) and ide2 (empty). The VM that I deployed only had one CD-ROM drive: ide2 (cloudinit)

I know people on Proxmox 8.1 can't downgrade to 8.0 to test, but if there's someone who has a test server running 8.0, it'd be nice to test this against that version as well. I haven't seen anyone report this problem with 8.1, but I have seen people talk about Proxmox introducing some breaking changes in 8.1 (things that worked fine with our provider in 8.0). I don't want to upgrade my test server for this same reason: I can't easily go back to 7.4 if I upgrade.

Side note: In case anyone is wondering why my base VM had two IDE drives, I'm right there with you. Apparently that's something that packer did when I initially created that image. :shrug:

**Comment by devZer0** [comment]
> I have been having this issue as well. The solution I finally found to work was similar to hestiahacker's #5. I removed the cloud-init drive from the template. The difference or at least what wasn't stated was that in order to add it back I just specified `cloudinit_cdrom_storage = "local-zfs"`. This created a new cloud-init drive on the cloned VM that then assigned its own slot. This also retained the cloud-init settings that were being used by the template before I removed the drive.

this param (and also this one) also fixed the problem for me with pve 8.1.3

**Comment by spensireli** [comment]
Still seems to be apparent in v3.0.1-rc1. Terraform clones the template and it initially looks great then for some reason removes the cloud-init drive?

**Comment by n8rade** [comment]
> Still seems to be apparent in v3.0.1-rc1. Terraform clones the template and it initially looks great then for some reason removes the cloud-init drive?

I don't believe that would be related to this issue and that may be the fault of my generic title. As far as the specifics of this issue goes unless you can share logs I believe you may be running into something like is mentioned in #915 

If memory serves I ran into the same issue switching over to 3.0.1-rc1. Lots of changes!

**Comment by n8rade** [comment]
At this point I've been using v3.0.1-rc1 for awhile now without this issue. I'm assuming this is fixed and will be fixed in the full release as well. Closing this out, thank you!!
